app/internal/features.py

```diff
@@ -0,0 +1,183 @@
+) -> None:
+    session.query(UserFeature).filter_by(feature_id=feature.id).delete()
+    session.query(Feature).filter_by(id=feature.id).delete()
+    session.commit()
+
+
+def is_feature_exists(feature: Dict[str, str], session: Session) -> bool:
+    is_exists = session.query(
+        exists()
+        .where(Feature.name == feature["name"])
+        .where(Feature.route == feature["route"]),
+    ).scalar()
+
+    return is_exists
+
+
+def update_feature(
+    feature: Feature,
+    feature_dict: Dict[str, str],
+    session: Session = Depends(get_db),
+) -> Feature:
+    feature.name = feature_dict["name"]
+    feature.route = feature_dict["route"]
+    feature.description = feature_dict["description"]
+    feature.creator = feature_dict["creator"]
+    session.commit()
+    return feature
+
+
+async def is_access_allowd(request: Request, route: str) -> bool:
+    session = SessionLocal()
+
+    # Get current user.
+    # Note: can't use dependency beacause its designed for routes only.
+    # current_user return schema not an db model.
+    jwt = await get_authorization_cookie(request=request)
+    user = await current_user(request=request, jwt=jwt, db=session)
+
+    feature = session.query(Feature).filter_by(route=route).first()
+
+    if feature is None:
+        # in case there is no feature exists in the database that match the
+        # route that gived by to the request.
+        return True
+
+    user_ptef = session.query(
+        exists().where(
+            (UserFeature.feature_id == feature.id)
+            & (UserFeature.user_id == user.user_id),
+        ),
+    ).scalar()
+
+    return user_ptef
+
+
+def create_feature(
+    db: Session,
+    name: str,
+    route: str,
+    description: str,
+    creator: str = None,
+) -> Feature:
+    """Creates a feature."""
+    return create_model(
+        db,
+        Feature,
+        name=name,
+        route=route,
+        creator=creator,
+        description=description,
+    )
+
+
+def create_user_feature_association(
+    db: Session,
+    feature_id: int,
+    user_id: int,
+    is_enable: bool,
+) -> UserFeature:
+    """Creates an association."""
+    return create_model(
+        db,
+        UserFeature,
+        user_id=user_id,
+        feature_id=feature_id,
+        is_enable=is_enable,
+    )
+
+
+def get_user_installed_features(
+    user_id: int,
+    session: Session = Depends(get_db),
+) -> List[Feature]:
+    return (
+        session.query(Feature)
+        .join(UserFeature)
+        .filter(UserFeature.user_id == user_id)
+        .all()
+    )
+
+
+def get_user_uninstalled_features(
+    user_id: int,
+    session: Session = Depends(get_db),
+) -> List[Feature]:
+    return (
+        session.query(Feature)
+        .filter(
+            Feature.id.notin_(
+                session.query(UserFeature.feature_id).filter(
+                    UserFeature.user_id == user_id,
+                ),
+            ),
+        )
+        .all()
+    )
```

Comment:
That's a rather complicated way to do JOIN, isn't it? :)

Reply:
Couldn't do this with join...

Reply:
What to do?

Reply:
Why?

Reply:
Is there something more specific other than `join` you can refer me to?

Reply:
Nope. You can read about JOIN in the course notebooks

Reply:
> Why?

because there is multiple users and i need to query by their `user_id`.

Reply:
That's the purpose of `JOIN`. Please re-read about it.

Reply:
I'm working on it for 4 hours non stop, can't seem to make it work with join....
i've talk with 3 other people and they tried to help me and they can't seem to make it work either way!

Can you please help me? 
it's a many to many relationship and i need to get all the rows that don't have an association while making sure the association `user_id` equals to the parameter `user_id`, so it won't take other users data.

I couldn't do this with join, so i did it in query in side query...

Reply:
I talk to you on Telegram you approved this.